src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.DataFlowOperationVisitor.cs

```diff
@@ -0,0 +1,146 @@
+        {
+            private readonly WellKnownSymbolsInfo _wellKnownSymbolsInfo;
+            private readonly IMethodSymbol? _getEnumeratorMethod;
+
+            protected AvoidMultipleEnumerationsFlowStateDictionaryFlowOperationVisitor(
+                GlobalFlowStateDictionaryAnalysisContext analysisContext,
+                WellKnownSymbolsInfo wellKnownSymbolsInfo,
+                IMethodSymbol? getEnumeratorMethod) : base(analysisContext)
+            {
+                _wellKnownSymbolsInfo = wellKnownSymbolsInfo;
+                _getEnumeratorMethod = getEnumeratorMethod;
+            }
+
+            protected abstract bool IsExpressionOfForEachStatement(SyntaxNode node);
+
+            public override GlobalFlowStateDictionaryAnalysisValue VisitParameterReference(IParameterReferenceOperation operation, object? argument)
+            {
+                var value = base.VisitParameterReference(operation, argument);
+                if (!IsDeferredType(operation.Parameter.Type.OriginalDefinition, _wellKnownSymbolsInfo.AdditionalDeferredTypes))
+                {
+                    return value;
+                }
+
+                if (!IsOperationEnumeratedByMethodInvocation(operation, _wellKnownSymbolsInfo) && !IsGetEnumeratorOfForEachLoopInvoked(operation))
+                {
+                    return value;
+                }
+
+                if (!TryGetInvocationEntity(operation, out var invocationEntity))
+                {
+                    return value;
+                }
+
+                var newValue = CreateAnalysisValue(invocationEntity, operation, value);
+                UpdateGlobalValue(newValue);
+                return newValue;
+            }
+
+            public override GlobalFlowStateDictionaryAnalysisValue VisitLocalReference(ILocalReferenceOperation operation, object? argument)
+            {
+                var value = base.VisitLocalReference(operation, argument);
+                if (!IsDeferredType(operation.Local.Type.OriginalDefinition, _wellKnownSymbolsInfo.AdditionalDeferredTypes))
+                {
+                    return value;
+                }
+
+                if (!IsOperationEnumeratedByMethodInvocation(operation, _wellKnownSymbolsInfo) && !IsGetEnumeratorOfForEachLoopInvoked(operation))
+                {
+                    return value;
+                }
+
+                if (!TryGetInvocationEntity(operation, out var invocationEntity))
+                {
+                    return value;
+                }
+
+                var newValue = CreateAnalysisValue(invocationEntity, operation, value);
+                UpdateGlobalValue(newValue);
+                return newValue;
+            }
+
+            private bool TryGetInvocationEntity(IOperation operation, [NotNullWhen(true)] out InvocationEntity? invocationEntity)
+            {
+                var pointToAnalysisResult = DataFlowAnalysisContext.PointsToAnalysisResult;
```

Comment:
PointsToAnalysis has been known to be quite expensive. If you are planning to enable this rule by default, we should avoid relying on points to analysis.

Reply:
Marked this as resolved because in other comments there is a benchmark result